### ISSUE TITLE
MNT: Move to new numpy.random API in `test_*.py`

### ISSUE THREAD

**Issue by neutrinoceros** [astropy/astropy#17026]
Similarly to #16924, this will track progress for the re-issuing of #16086 over multiple smaller PRs, which I'll link here.
The end goal is to address https://github.com/astropy/astropy/issues/11687

**Comment by eerovaher** [comment]
It is not at all clear to me why we would ever want to generate random numbers with `numpy` in our tests. For unit tests it would be better to hard-code the values. [Property-based tests should be implemented with `hypothesis`.](https://docs.astropy.org/en/latest/development/testguide.html#property-based-tests)

**Comment by pllim** [comment]
Probably was copying from existing research code at the time or wanted convenience. I agree it is not strictly necessary but getting rid of it completely might be out of scope?

**Comment by eerovaher** [comment]
If something's being updated, it should be updated properly.

**Comment by neutrinoceros** [comment]
Well I see your point but:
- It is a lot more work than what I'm setting up to here, and I don't feel comfortable investing such an amount of my paid time in a large refactor (when I'm supposed to focus on bug fixes) without at least a large consensus being reached first.
- I've seen some reluctance against existing `hypothesis` tests in astropy because they're typically much slower to run, so I *definitely* will not pursue that track without a larger discussion first.

Overall I think it's out of scope if not off-topic, and would be better discussed in https://github.com/astropy/astropy/issues/11687 directly.
My immediate goal is just to avoid Mridul's (paid) work to go waste.

**Comment by eerovaher** [comment]
> Well I see your point but:
> 
> * It is a lot more work than what I'm setting up to here, and I don't feel comfortable investing such an amount of my paid time in a large refactor (when I'm supposed to focus on bug fixes) without at least a large consensus being reached first.

The detailed CI logs for the last commit in #16669 are no longer available, but it is still visible that it failed the CI. If you want to revive that pull request then you will have to make changes to it, and if you are making changes then you might as well make them properly.

> * I've seen some reluctance against existing `hypothesis` tests in astropy because they're typically much slower to run, so I _definitely_ will not pursue that track without a larger discussion first.

The advantage of using [Hypothesis](https://hypothesis.readthedocs.io/en/latest/) is that if it does detect a test failure then [the failure can be reproduced](https://hypothesis.readthedocs.io/en/latest/database.html) and Hypothesis will also [shrink](https://hypothesis.readthedocs.io/en/latest/data.html#shrinking) it, which is helpful for understanding the cause of the failure. Generating random numbers in a haphazard way does not offer these benefits.

> Overall I think it's out of scope if not off-topic, and would be better discussed in #11687 directly.

I am fully convinced that replacing RNG with hardcoded values would improve our unit tests and replacing our use of `numpy` RNG with Hypothesis would improve our property-based tests. Tweaking the details of `numpy` RNG in our tests is a waste of effort and pointing it out in a issue about how to use `numpy` RNG in our tests is not off-topic. #11687 remains relevant no matter how the discussion here concludes because that issue is not limited to tests.

**Comment by pllim** [comment]
Hypothesis was not the intend of the original PR also, so please move that part of convo to https://github.com/astropy/astropy/issues/9017 . Thanks!

**Comment by pllim** [comment]
If people are really opposed to straight replacing one random generator with another here, we have 2 options:

1. Let the original PR die. Walk away.
2. Change the scope to getting rid of random generator in tests (within reason). This might need discussions on `astropy-dev` first before actual work begin.

**Comment by eerovaher** [comment]
We have finite developer capacity, so I agree that it would be acceptable to leave the tests as they are, or to only remove random number generation from the tests that don't need them.

**Comment by pllim** [comment]
@neutrinoceros , do you have a preference on how you want to proceed given the limits of your contract? Thanks!

**Comment by neutrinoceros** [comment]
What I set out to do originally (resurect Mridul's work) doesn't seem like an expensive task, and following @eerovaher's suggestions on #17027, I now think it's also reasonable to take the opportunity to refactor rng *out* of certain tests where it's particularly easy.
My preference would be to start cherry-picking bits from https://github.com/astropy/astropy/pull/16086, but for the person doing it (me) to pay special attention on this aspect *before* opening PRs for review. How does this sound ?

**Comment by mhvk** [comment]
Agreed with @eerovaher here - we should really not have tests using random numbers in the first place (though with a fixed seed it is not horrible), and I don't see a big advantage in moving to the new API when we do - numpy has shown no intent of removing the legacy wrappers. It seems a lot of churn for little benefit (a general problem with these types of search-and-replace operations - an easy way to create a PR, not a particularly useful way to get to better code).

I'm also one of the ones a bit wary of moving to `hypothesis`. I love the principle and was excited when we implemented it for `Time`, but found the practice rather painful: a lot of warnings about rather irrelevant corner case (e.g., very big jd1 and jd2 of opposite sign). Of course, nice that those ended up being solved, but it meant days (weeks?) of my time that arguably could have been used better, plus now using an order of magnitude more time/cycles/energy for those tests every time CI is run. I guess as always one should consider the opportunity cost... In this case, at a minimum, I'd rather have a commitment from a maintainer sufficiently excited about hypothesis that they will maintain this for >=years.

Overall, I think letting the original PR die is not too bad, this seems low priority. But if it helps to identify poorly written tests and improves those, that may still be worthwhile. I just wonder if one isn't better off to try to bring a particular set of tests up to snuff...

**Comment by pllim** [comment]
> not have tests using random numbers in the first place

Well, if you want like thousands of data points that are not zeroes or ones, hardcoding or reading from file every run are not that fun either, compared to random generator with fixed seed. 🤷‍♀️
